### PR TITLE
Fix PathTranslator to compare path prefixes by path components

### PR DESCRIPTION
**Pull request by dbw9580**
### What changes are proposed in this pull request?

Previously, `PathTranslator` erroneously:

1. finds a prefix of `ufsPath` by simply checking with `String.startsWith()`. This fails to handle the case where the boundary falls within a path component. This case is now covered in the `prefixBoundaryWithinPathComponent` test;
2. chooses an arbitary prefix of `ufsPath` (the first returned during map iteration) even if there are multiple eligible choices. This is fixed by choosing the deepest prefix instead.

### Why are the changes needed?

It's a bug fix.

### Does this PR introduce any user facing changes?

No.

### CODE REVIEW

File: table/server/common/src/test/java/alluxio/table/common/udb/PathTranslatorTest.java [hunk 1]
```diff
@@ -76,24 +76,24 @@ public void multipleSubdir() throws Exception {
 
   @Test
   public void samePathDifferentUfs() throws Exception {
-    String alluxioPath  = "alluxio:///my/table/directory";
+    String alluxioPath  = "alluxio://master/my/table/directory";
     String ufsPath = "/the/ufs/location";
-    mTranslator.addMapping(alluxioPath, "ufs_a://" + ufsPath);
+    mTranslator.addMapping(alluxioPath, "ufs-a://" + ufsPath);
```
Comment (yuzhu) [inline table/server/common/src/test/java/alluxio/table/common/udb/PathTranslatorTest.java:81]: why was this changed? does '_' not work?

File: table/server/common/src/test/java/alluxio/table/common/udb/PathTranslatorTest.java [hunk 2]
```diff
@@ -76,24 +76,24 @@ public void multipleSubdir() throws Exception {
 
   @Test
   public void samePathDifferentUfs() throws Exception {
-    String alluxioPath  = "alluxio:///my/table/directory";
+    String alluxioPath  = "alluxio://master/my/table/directory";
```
Comment (yuzhu) [inline table/server/common/src/test/java/alluxio/table/common/udb/PathTranslatorTest.java:79]: we should also be able to support path like 'alluxio:///dir/file' does this change not work with that kind of path?

File: table/server/common/src/main/java/alluxio/table/common/udb/PathTranslator.java [hunk 3]
```diff
@@ -45,16 +48,7 @@ public PathTranslator() {
...
-      // strip trailing slashes
-      alluxioPath = alluxioPath.substring(0, alluxioPath.length() - 1);
-    }
-
-    while (ufsPath.endsWith("/")) {
-      // strip trailing slashes
-      ufsPath = ufsPath.substring(0, ufsPath.length() - 1);
-    }
-    mPathMap.put(alluxioPath, ufsPath);
+    mPathMap.put(new AlluxioURI(alluxioPath), new AlluxioURI(ufsPath));
```
Comment (yuzhu) [inline table/server/common/src/main/java/alluxio/table/common/udb/PathTranslator.java:51]: what is the reason to move this to the translation time? would this add potentially duplicate keys like `/path1/path2' and `/path1/path2/` ?

**Review by yuzhu — COMMENTED** [review]
@dbw9580 Thanks for the fix! good catch, have some questions about adding path and tests.

File: table/server/common/src/test/java/alluxio/table/common/udb/PathTranslatorTest.java (hunk 1 above)
Comment (dbw9580) [reply table/server/common/src/test/java/alluxio/table/common/udb/PathTranslatorTest.java:81]: `java.lang.IllegalArgumentException: java.net.URISyntaxException: Illegal character in scheme name at index 3: ufs_a:/the/ufs/location`

It seems "_" is forbidden in schemes.

File: table/server/common/src/test/java/alluxio/table/common/udb/PathTranslatorTest.java (hunk 2 above)
Comment (dbw9580) [reply table/server/common/src/test/java/alluxio/table/common/udb/PathTranslatorTest.java:79]: https://github.com/Alluxio/alluxio/blob/90b3c7b010f926be3c2df7519eec067df30fd2a7/table/server/common/src/main/java/alluxio/table/common/udb/PathTranslator.java#L96-100

I check for the existence of both scheme and authority. Is this too strict?

File: table/server/common/src/main/java/alluxio/table/common/udb/PathTranslator.java (hunk 3 above)
Comment (dbw9580) [reply table/server/common/src/main/java/alluxio/table/common/udb/PathTranslator.java:51]: > move this to the translation time

Do you mean the creation of `AlluxioURI`s? I simply use them for the convenient methods like `isAncestorOf` and that it trims trailing slashes already.

File: table/server/common/src/test/java/alluxio/table/common/udb/PathTranslatorTest.java (hunk 2 above)
Comment (yuzhu) [reply table/server/common/src/test/java/alluxio/table/common/udb/PathTranslatorTest.java:79]: I think it is ok to not have authority, specifically i think for HA URIs. see https://docs.alluxio.io/os/user/stable/en/deploy/Running-Alluxio-On-a-HA-Cluster.html#specify-alluxio-service-in-configuration-parameters-or-java-options

File: table/server/common/src/main/java/alluxio/table/common/udb/PathTranslator.java (hunk 3 above)
Comment (yuzhu) [reply table/server/common/src/main/java/alluxio/table/common/udb/PathTranslator.java:51]: oh somehow i missed that you create AlluxioURIs out of them, yeah this works

File: table/server/common/src/test/java/alluxio/table/common/udb/PathTranslatorTest.java (hunk 1 above)
Comment (yuzhu) [reply table/server/common/src/test/java/alluxio/table/common/udb/PathTranslatorTest.java:81]: yup, it makes sense now that I see you create AlluxioURI out of the path

**Review by yuzhu — APPROVED** [review]
LGTM once the test passes and URI without authority is addressed

File: table/server/common/src/test/java/alluxio/table/common/udb/PathTranslatorTest.java (hunk 2 above)
Comment (dbw9580) [reply table/server/common/src/test/java/alluxio/table/common/udb/PathTranslatorTest.java]: Changed in 825c0cc.

**Comment by dbw9580** [comment]
@yuzhu forgot to add test for the deepest match part. PTAL, thanks.

**Review by yuzhu — APPROVED** [review]

**Comment by yuzhu** [comment]
alluxio-bot, merge this please
